File: Java/820.java
__________________________________________________________________________________________________
sample 5 ms submission
class Solution {
    public class Trie{
        Trie[] children;
        int count;
        public Trie(){
            children = new Trie[26];
            count = 0;
        }
    }
    Trie root;
    int res = 0;
    public int minimumLengthEncoding(String[] words) {
        root = new Trie();
        for(String word:words){
            add(word);
        }
        return res;
    }
    public void add(String word){
        Trie cur = root;
        boolean flag = false;
        for(int i = word.length()-1; i >= 0; i--){
            char c = word.charAt(i);
            if(cur.children[c-'a'] == null){
                cur.children[c-'a'] = new Trie();
                flag = true;
            }
            cur = cur.children[c-'a'];
            if(cur.count > 0 && i != 0){
                res -= cur.count+1;
                cur.count = 0;
            }
        }
        if(flag == false) return;
        cur.count = word.length();
        res += word.length()+1;
    }
}
__________________________________________________________________________________________________
sample 38328 kb submission
class Solution {
    public int minimumLengthEncoding(String[] words) {
        Node root = new Node();
        int result = 0;
        Arrays.sort(words, (o1, o2) -> Integer.compare(o2.length(), o1.length()));
        for (String word : words) {
            Node cur = root;
            boolean first = false;
            for (int i = word.length() - 1; i >= 0; --i) {
                char c = word.charAt(i);
                if (cur.child[c - 'a'] == null) {
                    cur.child[c - 'a'] = new Node();
                    first = true;
                }
                cur = cur.child[c - 'a']; 
            }
            if (first == true) {
                result += word.length() + 1;
            }
            
        }
        
        return result;   
    }
    
    class Node {
        Node[] child = new Node[26];
    }
}
__________________________________________________________________________________________________

File: Python3/820.py
__________________________________________________________________________________________________
sample 100 ms submission
class Solution:
    def minimumLengthEncoding(self, words: List[str]) -> int:
        d = {}
        for word in words:
            n = len(word)
            if n in d:
                d[n].add(word)
            else:
                d[n]  = set([word])
        keys = list(d.keys())
        keys.sort()
        n = len(keys)
        res = 0
        for i in range(n-1, -1, -1):
            if i == n-1:
                s = d[keys[i]]
                L = len(d[keys[i]])
                res += L * keys[i] + L
            else:  
                s = set([item[-keys[i]:] for item in s])
                L1 = len(s)
                s = s | d[keys[i]]
                L3 = len(s)
                res += (L3 - L1) * keys[i] + (L3 - L1)
        return res        
__________________________________________________________________________________________________
sample 15540 kb submission
from functools import reduce
class Solution:
    def minimumLengthEncoding(self, words: List[str]) -> int:
        T = lambda: collections.defaultdict(T)
        root = T()
        end = []
        for w in set(words):
            end.append(reduce(lambda x,y: x[y], w[::-1], root) )
            end[-1]['#'] = len(w)+1
        
        return sum(w['#'] for w in end if len(w)==1)
__________________________________________________________________________________________________
